sys/net/gnrc/link_layer/lwmac/lwmac.c

```diff
@@ -0,0 +1,852 @@
+            /* Quit scheduling transmission if 'quit-tx' flag is found set, thus
+             * to avoid potential collisions with ongoing transmissions of other
+             * neighbor nodes */
+            if (gnrc_netdev_lwmac_get_quit_tx(gnrc_netdev)) {
+                return false;
+            }
+
+            _sleep_management(gnrc_netdev);
+            break;
+        }
+        case LWMAC_LISTENING: {
+            /* In case has pending packet to send, clear rtt alarm thus to goto
+             * transmission initialization (in SLEEPING management) right after the
+             * listening period */
+            if ((_next_tx_neighbor(gnrc_netdev) != NULL) ||
+                (gnrc_netdev->tx.current_neighbor != NULL)) {
+                rtt_handler(LWMAC_EVENT_RTT_PAUSE, gnrc_netdev);
+            }
+
+            /* Set timeout for if there's no successful rx transaction that will
+             * change state to SLEEPING. */
+            if (!lwmac_timeout_is_running(gnrc_netdev, TIMEOUT_WAKEUP_PERIOD)) {
+                lwmac_set_timeout(gnrc_netdev, TIMEOUT_WAKEUP_PERIOD, LWMAC_WAKEUP_DURATION_US);
+            }
+            else if (lwmac_timeout_is_expired(gnrc_netdev, TIMEOUT_WAKEUP_PERIOD)) {
+                /* Dispatch first as there still may be broadcast packets. */
+                _dispatch(gnrc_netdev->rx.dispatch_buffer);
+
+                gnrc_netdev->lwmac.state = LWMAC_SLEEPING;
+                /* Enable duty cycling again */
+                rtt_handler(LWMAC_EVENT_RTT_RESUME, gnrc_netdev);
+
+                _set_netdev_state(gnrc_netdev, NETOPT_STATE_SLEEP);
+                lwmac_clear_timeout(gnrc_netdev, TIMEOUT_WAKEUP_PERIOD);
+
+                /* if there is a packet for transmission, schedule update to start
+                 * transmission initialization immediately. */
+                gnrc_mac_tx_neighbor_t *neighbour = _next_tx_neighbor(gnrc_netdev);
+                if ((neighbour != NULL) || (gnrc_netdev->tx.current_neighbor != NULL)) {
+                    /* This triggers packet sending procedure in sleeping immediately. */
+                    lwmac_schedule_update(gnrc_netdev);
+                    break;
+                }
+            }
+
+            if (gnrc_priority_pktqueue_length(&gnrc_netdev->rx.queue) > 0) {
+                /* Do wake-up extension in each packet reception. */
+                lwmac_clear_timeout(gnrc_netdev, TIMEOUT_WAKEUP_PERIOD);
+                lwmac_set_state(gnrc_netdev, LWMAC_RECEIVING);
+            }
+            break;
+        }
+        case LWMAC_RECEIVING: {
+            _rx_management(gnrc_netdev);
+            break;
+        }
+        case LWMAC_TRANSMITTING: {
+            _tx_management(gnrc_netdev);
+            break;
+        }
+        default:
+            LOG_DEBUG("No actions in state %u\n", gnrc_netdev->lwmac.state);
+    }
+
+    return gnrc_netdev_lwmac_get_reschedule(gnrc_netdev);
+}
+
+static void rtt_cb(void *arg)
+{
+    msg_t msg;
```

Comment:
This goes out of scope as soon as the ISR returns, before the other thread has handled it. This might be the source of the random hardfaults mentioned somewhere in the comments.